compiler/vm/src/builtin_functions.rs

```diff
@@ -170,48 +150,46 @@ macro_rules! unpack_and_later_drop {
     };
 }
 
-impl Heap {
-    fn channel_create(&mut self, args: &[InlineObject]) -> BuiltinResult {
-        unpack_and_later_drop!(self, args, |capacity: Int| {
-            let capacity = capacity.try_get().unwrap();
-            CreateChannel { capacity }
-        })
-    }
-    fn channel_send(&mut self, args: &[InlineObject]) -> BuiltinResult {
-        unpack_and_later_drop!(self, args, |port: SendPort, packet: Any| {
-            let mut heap = Heap::default();
-            let object = packet.object.clone_to_heap(&mut heap);
-            Send {
-                channel: port.channel_id(),
-                packet: Packet { heap, object },
-            }
-        })
-    }
-    fn channel_receive(&mut self, args: &[InlineObject]) -> BuiltinResult {
-        unpack_and_later_drop!(self, args, |port: ReceivePort| {
-            Receive {
-                channel: port.channel_id(),
-            }
-        })
-    }
+use SuccessfulBehavior::*;
 
+impl Heap {
     fn equals(&mut self, args: &[InlineObject]) -> BuiltinResult {
         unpack_and_later_drop!(self, args, |a: Any, b: Any| {
             Return(Tag::create_bool(**a == **b).into())
         })
     }
 
     fn function_run(&mut self, args: &[InlineObject], responsible: HirId) -> BuiltinResult {
-        unpack!(self, args, |function: Function| {
-            DivergeControlFlow {
-                function: *function,
-                responsible,
+        unpack!(self, args, |function: Any| {
+            match **function {
+                Data::Builtin(_) => {
+                    // TODO: Replace with `unreachable!()` once we have guards
+                    // for argument counts on the Candy side – there are no
+                    // builtins without arguments.
+                    return Err("`function.run` called with builtin".to_string());
+                }
+                Data::Function(function) => DivergeControlFlow {
+                    function,
+                    responsible,
+                },
+                Data::Handle(handle) => CallHandle(CallHandle {
+                    handle,
+                    arguments: vec![],
```

Comment:
Validation for the number of arguments is missing here and `machine_state.call(…)`